### PR TITLE
pkg/asset/tls: add etcd metrics assets

### PR DESCRIPTION
**Pull request by hexfusion**
This PR is to start a conversation about using etcd grpc-proxy to provide TLS trust separation between etcd server and monitoring. The idea here is to provide monitoring a root CA which they can own giving access to etcd metrics but not etcd server directly. etcd grpc-proxy will only forward GET requests to etcd server /metrics endpoint.

/cc @abhinavdahiya 

Blocking: https://github.com/openshift/machine-config-operator/pull/517.

### CODE REVIEW

File: pkg/asset/tls/etcdmetricsproxy.go [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+
+import (
+	"crypto/x509"
+	"crypto/x509/pkix"
+
+	"github.com/openshift/installer/pkg/asset"
+)
+
+// EtcdMetricsProxyCA is the asset that generates the etcd-metrics-proxy-ca key/cert pair.
+// [DEPRECATED]
```
Comment (abhinavdahiya) [inline pkg/asset/tls/etcdmetricsproxy.go]: a new asset is already deprecated? ;)

**Comment by abhinavdahiya** [comment]
we are moving towards dropping root-ca from all chains. some examples https://github.com/openshift/installer/pull/1232

File: pkg/asset/tls/etcdmetricsproxy.go (hunk 1 above)
Comment (hexfusion) [reply pkg/asset/tls/etcdmetricsproxy.go]: yeah started with your template but I see where you are going now will clean that up, thanks for the feedback!

**Comment by hexfusion** [comment]
@abhinavdahiya does this look more inline with what you are expecting?

File: pkg/asset/tls/etcdmetrics.go [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+}
+
+// Generate generates the cert/key pair based on its dependencies.
+func (a *EtcdMetricsSignerClientCertKey) Generate(dependencies asset.Parents) error {
+	ca := &EtcdMetricsSignerCertKey{}
+	dependencies.Get(ca)
+
+	cfg := &CertCfg{
+		Subject:      pkix.Name{CommonName: "etcd-metrics", OrganizationalUnit: []string{"etcd-metrics"}},
+		KeyUsages:    x509.KeyUsageKeyEncipherment,
```
Comment (mrogers950) [inline pkg/asset/tls/etcdmetrics.go]: Add x509.KeyUsageDigitalSignature as well

Comment (hexfusion) [reply pkg/asset/tls/etcdmetrics.go]: Updated

File: data/data/manifests/bootkube/etcd-service.yaml [hunk 3]
```diff
@@ -13,3 +13,6 @@ spec:
   - name: etcd
     port: 2379
     protocol: TCP
+  - name: etcd-metrics
+    port: 23790
+    protocol: TCP
```
Comment (hexfusion) [inline data/data/manifests/bootkube/etcd-service.yaml]: Should we drop this as a service for metrics? it really isn't useful they need to connect to each member explicitly. We will provide this via endpoint nodeName such as `https://ip-10-0-150-40.ec2.internal:23790`  @brancz ?

File: pkg/asset/tls/etcdmetrics.go [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+	cfg := &CertCfg{
+		Subject:      pkix.Name{CommonName: "etcd-metrics", OrganizationalUnit: []string{"etcd-metrics"}},
+		KeyUsages:    x509.KeyUsageKeyEncipherment | x509.KeyUsageDigitalSignature,
+		ExtKeyUsages: []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
+		DNSNames: []string{
+			"etcd",
+			"etcd.kube-system",
+			"etcd.kube-system.svc.cluster.local",
+			"etcd.kube-system.svc",
+			"*.ec2.internal",
```
Comment (hexfusion) [inline pkg/asset/tls/etcdmetrics.go]: not really optimal or cloud agnostic

File: data/data/manifests/bootkube/etcd-service.yaml (hunk 3 above)
Comment (hexfusion) [reply data/data/manifests/bootkube/etcd-service.yaml]: keeping

File: pkg/asset/tls/etcdmetrics.go (hunk 4 above)
Comment (hexfusion) [reply pkg/asset/tls/etcdmetrics.go]: removed

**Comment by hexfusion** [comment]
@abhinavdahiya PTAL

File: data/data/aws/vpc/sg-master.tf [hunk 5]
```diff
@@ -216,6 +216,16 @@ resource "aws_security_group_rule" "master_ingress_etcd" {
   self      = true
 }
 
+resource "aws_security_group_rule" "master_ingress_etcd_metrics" {
+  type              = "ingress"
+  security_group_id = "${aws_security_group.master.id}"
+
+  protocol  = "tcp"
+  from_port = 23790
```
Comment (abhinavdahiya) [inline data/data/aws/vpc/sg-master.tf]: can you get approval to use that port https://github.com/openshift/origin/pull/21520

File: pkg/asset/manifests/operators.go [hunk 6]
```diff
@@ -145,6 +153,9 @@ func (m *Manifests) generateBootKubeManifests(dependencies asset.Parents) []*ass
 		EtcdCaCert:                      string(etcdCA.Cert()),
 		EtcdClientCert:                  base64.StdEncoding.EncodeToString(etcdClientCertKey.Cert()),
 		EtcdClientKey:                   base64.StdEncoding.EncodeToString(etcdClientCertKey.Key()),
+		EtcdMetricsCaCert:               string(etcdMetricsSignerCertKey.Cert()),
```
Comment (abhinavdahiya) [inline pkg/asset/manifests/operators.go]: shouldn't you be using 
https://github.com/openshift/installer/pull/1291/files#diff-303f2bd7732f182add6f8ae14e9929f1R41

**Comment by abhinavdahiya** [comment]
@hexfusion this is LGTM.
- can you squash to more reasonable number of commits. (if you want to do one that's fine too.)
- https://github.com/openshift/origin/pull/21520 make sure `23790` is one of the recognized port.

File: pkg/asset/manifests/operators.go (hunk 6 above)
Comment (hexfusion) [reply pkg/asset/manifests/operators.go]: @abhinavdahiya I would like to keep it like this if possible. It feels like we could drop the CA Bundle config but maybe I am missing something?

**Comment by hexfusion** [comment]
@abhinavdahiya rebased commits. WRT https://github.com/openshift/origin/pull/21520 would it preferred to drop into the `9xxx` port range or just wait for verification. I have no strong feeling on the port.

File: pkg/asset/manifests/operators.go (hunk 6 above)
Comment (abhinavdahiya) [reply pkg/asset/manifests/operators.go]: this seems to used for https://github.com/openshift/installer/pull/1291/files#diff-626d6cfa425c30cb6c099ebc5324d25b
I think we should use the CABundle asset for this for consistency.

Comment (hexfusion) [reply pkg/asset/manifests/operators.go]: ack will adjust

**Comment by hexfusion** [comment]
```
[root@ip-10-0-172-22 ~]# ss -ltn | grep -Eo ':[9][0-9]{3}' | sort --unique
:9100
:9101
:9537
```

**Comment by brancz** [comment]
Generally :+1: from me as not having etcd metrics is a regression from 3.11, and it's by far the most important component to be monitored. This must be in 4.0.

**Pull request merged**